### PR TITLE
 8319417: (dc) DatagramChannel.connect undocumented behavior

### PR DESCRIPTION
**Pull request by Michael-Mc-Mahon**
Hi,

This is a small doc change for DatagramChannel.getLocalAddress() and the equivalent methods on DatagramSocket which documents some existing (but not currently documented) behavior.

Thanks,
Michael

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Change requires CSR request [JDK-8319962](https://bugs.openjdk.org/browse/JDK-8319962) to be approved
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8319417](https://bugs.openjdk.org/browse/JDK-8319417): (dc) DatagramChannel.connect undocumented behavior (**Enhancement** - P4)
 * [JDK-8319962](https://bugs.openjdk.org/browse/JDK-8319962): (dc) DatagramChannel.connect undocumented behavior (**CSR**)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [0dfe2bf7](https://git.openjdk.org/jdk/pull/16635/files/0dfe2bf7ccb7d4f3b5794c3fb1f478fb29fa0c90)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16635/head:pull/16635` \
`$ git checkout pull/16635`

Update a local copy of the PR: \
`$ git checkout pull/16635` \
`$ git pull https://git.openjdk.org/jdk.git pull/16635/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16635`

View PR using the GUI difftool: \
`$ git pr show -t 16635`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16635.diff">https://git.openjdk.org/jdk/pull/16635.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16635#issuecomment-1829465323)

### CODE REVIEW

File: src/java.base/share/classes/java/nio/channels/DatagramChannel.java [hunk 1]
```diff
@@ -655,7 +662,7 @@ public final long write(ByteBuffer[] srcs) throws IOException {
      * @return  The {@code SocketAddress} that the socket is bound to, or the
      *          {@code SocketAddress} representing the loopback address if
      *          denied by the security manager, or {@code null} if the
-     *          channel's socket is not bound
+     *          channel's socket is not bound.
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/nio/channels/DatagramChannel.java:665]: The period at the end of the return description isn't needed, the rest looks good.

**Comment by msheppar** [comment]
should not the lead sentence for this method be: 
Returns the local address to which the DatagramSocket or DatagramChannel has been bound.

Also, the current wording says that:
"If the channel's socket was initially bound to the wildcard address and
is now {@link #isConnected connected}, then the address returned
may be the local address selected as the source address for
datagrams sent via this channel instead of the wildcard address."

The may is a bit ambiguous ... is this OS dependent or is it universal that for a connected DatagramSocket or DatagramChannel then the local address of the intreface for which the send was routed will be returned. If it is OS dependent, should that be stated, also ?

Also, the return for the case of an unbound DatagramSocket or DatagramChannel is not specified.

**Comment by Michael-Mc-Mahon** [comment]
> The current wording says that: "If the channel's socket was initially bound to the wildcard address and is now {@link #isConnected connected}, then the address returned may be the local address selected as the source address for datagrams sent via this channel instead of the wildcard address."
> 
> The may is a bit ambiguous ... is this OS dependent or is it universal that for a connected DatagramSocket or DatagramChannel then the local address of the intreface for which the send was routed will be returned. If it is OS dependent, should that be stated, also ?
> 
> Also, the return for the case of an unbound DatagramSocket or DatagramChannel is not specified.

That is the behavior for the main supported platforms. We had a concern that it might not be the case on other platforms. Hence, the word "may".

**Comment by msheppar** [comment]
> That is the behavior for the main supported platforms. We had a concern that it might not be the case on other platforms. Hence, the word "may".

yes, appreciate that sentiment, but  "may" is a bit flippy floppy - a definitely maybe. A stronger emphasis that this OS dependent, such as the inclusion of an perhaps an apiNote to that effect,  would be useful and informative

**Comment by msheppar** [comment]
> > Also, the return for the case of an unbound DatagramSocket or DatagramChannel is not specified.
> 
> DatagramChannel::getLocalAddress is specified to return null when channel's socket is not bound. DatagramSocket::getLocalSocketAddress also specifies that it returns null when not bound. But maybe you mean something else?

Yes, indeed I did mean something else, I seemed to have lost the tail end of my sentence (the punchline, so to say)
It should have read something like:

Also, the return for the case of an unbound DatagramSocket or DatagramChannel is not specified consistently, and provide conflicting return specifications.

DatagramChannel return spec

Returns:
    The SocketAddress that the socket is bound to, or the SocketAddress representing the loopback address if denied by the security manager, or null if the channel's socket is not bound.

DatagramSocket return spec

Returns:
    the local address to which the socket is bound, null if the socket is closed, or an InetAddress representing [wildcard](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/net/InetAddress.html#isAnyLocalAddress()) address if either the socket is not bound, or the security manager checkConnect method does not allow the operation

**Comment by AlanBateman** [comment]
> Returns: the local address to which the socket is bound, null if the socket is closed, or an InetAddress representing [wildcard](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/net/InetAddress.html#isAnyLocalAddress()) address if either the socket is not bound, or the security manager checkConnect method does not allow the operation

DatagramSocket.getLocalAddress dates from JDK 1.1 and its odd behavior couldn't be changed for compatibility reasons. The newer DatagramSocket.getLocalSocketAddress is specified to return null when not bound.

**Comment by msheppar** [comment]
> > Returns: the local address to which the socket is bound, null if the socket is closed, or an InetAddress representing [wildcard](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/net/InetAddress.html#isAnyLocalAddress()) address if either the socket is not bound, or the security manager checkConnect method does not allow the operation
> 
> DatagramSocket.getLocalAddress dates from JDK 1.1 and its odd behavior couldn't be changed for compatibility reasons. The newer DatagramSocket.getLocalSocketAddress is specified to return null when not bound.

Hmm! sound like a bit of a fudge, and doesn't really make sense, as DatagramSocket and DatagramChannel essentially delegate to the same underlying implementation, which in turn delgates to the same underlying native implementation for which the OS will have a strong influence on the semantics offered at the higher level. 
Thus, in the case of an unbounded lower level socket, upon which a send (or connect) has been invoked, a subsequent getsockname would return whatever the OS has implicitly bound to the socket in terms of address and port. 
Thus, a similar scenario would be expected at the Java API level, and should result in the getLocalAddress method exhibiting the same semantics, returning  the same InetAddress for all specified conditions, irrespective of which API is called.
If both APIs delegate to the same underlying implementation, then it is reasonable to expect that they both exhibit a consistent samantics in their APIs. This is  especially true as DatagramChannel  exposes an associated DatagramSocket, which based your statement  will allow  a duality of behaviour to exist.
Regardless of compatibility, this is reasonably conceived as broken behaviour.

**Comment by AlanBateman** [comment]
> Hmm! sound like a bit of a fudge, and doesn't really make sense, as DatagramSocket and DatagramChannel essentially delegate to the same underlying implementation, which in turn delgates to the same underlying native implementation for which the OS will have a strong influence on the semantics offered at the higher level. Thus, in the case of an unbounded lower level socket, upon which a send (or connect) has been invoked, a subsequent getsockname would return whatever the OS has implicitly bound to the socket in terms of address and port. Thus, a similar scenario would be expected at the Java API level, and should result in the getLocalAddress method exhibiting the same semantics, returning the same InetAddress for all specified conditions, irrespective of which API is called. If both APIs delegate to the same underlying implementation, then it is reasonable to expect that they both exhibit a consistent samantics in their APIs. This is especially true as DatagramChannel exposes an associated DatagramSocket, which based your statement will allow a duality of behaviour to exist. Regardless of compatibility, this is reasonably conceived as broken behaviour.

DatagramSocket::getLocalAddress dates from JDK 1.0 and originally specified/implemented to return an InetAddress. The support for unbound sockets was added in JDK 1.4 (2004) and it would have been an incompatible change to change this method to return null, hence the current spec. I agree it's an odd API for this case but changing it now after all this time would be an incompatible change.

**Comment by Michael-Mc-Mahon** [comment]
> > That is the behavior for the main supported platforms. We had a concern that it might not be the case on other platforms. Hence, the word "may".
> 
> yes, appreciate that sentiment, but "may" is a bit flippy floppy - a definitely maybe. A stronger emphasis that this is OS dependent, such as the inclusion of an perhaps an apiNote to that effect, would be useful and informative

I think it's the same effect. I'd prefer not to complicate the text. We don't actually know the exact conditions where it might be supported and when not.

**Pull request closed without merging**